Flaky-test: [SimpleIntegrationTest].[testSubscribeRejectionWithSameClientId]

[SimpleIntegrationTest] is flaky. The [testSubscribeRejectionWithSameClientId] test method fails sporadically.

[example failure](https://github.com/streamnative/mop/runs/5869666283?check_suite_focus=true#logs)

```
Error:  testSubscribeRejectionWithSameClientId(io.streamnative.pulsar.handlers.mqtt.mqtt3.fusesource.base.SimpleIntegrationTest)  Time elapsed: 0.068 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
```

<details>
<summary>Full exception stacktrace</summary>
<code><pre>
Error:  Tests run: 50, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 44.629 s <<< FAILURE! - in io.streamnative.pulsar.handlers.mqtt.mqtt3.fusesource.base.SimpleIntegrationTest
Error:  testSubscribeRejectionWithSameClientId(io.streamnative.pulsar.handlers.mqtt.mqtt3.fusesource.base.SimpleIntegrationTest)  Time elapsed: 0.068 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.failNotEquals(Assert.java:776)
	at org.testng.Assert.assertTrue(Assert.java:44)
	at org.testng.Assert.assertTrue(Assert.java:54)
	at io.streamnative.pulsar.handlers.mqtt.mqtt3.fusesource.base.SimpleIntegrationTest.testSubscribeRejectionWithSameClientId(SimpleIntegrationTest.java:289)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

[INFO] 
</pre></code>
</details>

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/Connection.java
@@ -145,28 +145,21 @@ public class Connection {
         return channel.close();
     }
 
-    public CompletableFuture<Void> close() {
-        return close(false)
-                .exceptionally(ex -> {
-                    log.error("close connection : {} error", this, ex);
-                    return null;
-                });
+    public void disconnect() {
+        if (MqttUtils.isMqtt5(protocolVersion)) {
+            MqttMessage mqttMessage = MqttMessageBuilders
+                    .disconnect()
+                    .reasonCode(Mqtt5DisConnReasonCode.SESSION_TAKEN_OVER.byteValue())
+                    .build();
+            sendThenClose(mqttMessage);
+        } else {
+            channel.close();
+        }
     }
 
-    public CompletableFuture<Void> close(boolean force) {
-        log.info("Closing connection clientId = {} force : {}", clientId, force);
-        if (force) {
-            assignState(ESTABLISHED, DISCONNECTED);
-            if (MqttUtils.isMqtt5(protocolVersion)) {
-                MqttMessage mqttMessage = MqttMessageBuilders
-                        .disconnect()
-                        .reasonCode(Mqtt5DisConnReasonCode.SESSION_TAKEN_OVER.byteValue())
-                        .build();
-                sendThenClose(mqttMessage);
-            } else {
-                channel.close();
-            }
-        }
+    public CompletableFuture<Void> close() {
+        log.info("Closing connection clientId = {}", clientId);
+        assignState(ESTABLISHED, DISCONNECTED);
         // unregister all listener
         for (PulsarEventListener listener : listeners) {
             eventCenter.unRegister(listener);

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTConnectionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTConnectionManager.java
@@ -55,11 +55,7 @@ public class MQTTConnectionManager {
         Connection existing = connections.put(connection.getClientId(), connection);
         if (existing != null) {
             log.warn("The clientId is existed. Close existing connection. CId={}", existing.getClientId());
-            existing.close(true)
-                    .exceptionally(ex -> {
-                        log.error("close existing connection : {} error", existing, ex);
-                        return null;
-                    });
+            existing.disconnect();
         }
     }
 
@@ -93,7 +89,7 @@ public class MQTTConnectionManager {
     }
 
     public void close() {
-        connections.values().forEach(connection -> connection.close());
+        connections.values().forEach(connection -> connection.getChannel().close());
     }
 
     class ConnectEventListener implements EventListener {
@@ -106,12 +102,7 @@ public class MQTTConnectionManager {
                     Connection connection = getConnection(connectEvent.getClientId());
                     if (connection != null) {
                         log.warn("[ConnectEvent] close existing connection : {}", connection);
-                        connection.close(true)
-                                .thenRun(() -> removeConnection(connection))
-                                .exceptionally(ex -> {
-                                    log.error("[ConnectEvent] close existing connection : {} error", connection, ex);
-                                    return null;
-                                });
+                        connection.disconnect();
                     }
                 }
             }

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -13,7 +13,6 @@
  */
 package io.streamnative.pulsar.handlers.mqtt;
 
-import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.Maps;
 import io.streamnative.pulsar.handlers.mqtt.exception.MQTTNoSubscriptionExistedException;
 import java.util.List;
@@ -21,6 +20,7 @@ import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections4.CollectionUtils;
 import org.apache.commons.lang3.tuple.Pair;
 import org.apache.pulsar.broker.service.BrokerServiceException;
 import org.apache.pulsar.broker.service.Consumer;
@@ -61,11 +61,17 @@ public class TopicSubscriptionManager {
             log.error("[ Subscription ] Subscription {} Remove consumer fail.", subscriberName, e);
             FutureUtil.failedFuture(e);
         }
-
-        if (cleanSubscription) {
-            return subscriptionConsumerPair.getLeft().delete().thenAccept(__ -> topicSubscriptions.remove(topic));
+        if (CollectionUtils.isEmpty(subscriptionConsumerPair.getLeft().getConsumers())) {
+            return topic.unsubscribe(subscriberName)
+                    .thenCompose(__ -> {
+                        if (cleanSubscription) {
+                            return subscriptionConsumerPair.getLeft().deleteForcefully()
+                                    .thenAccept(unused -> topicSubscriptions.remove(topic));
+                        }
+                        return CompletableFuture.completedFuture(null);
+                    });
         } else {
-            return topic.unsubscribe(subscriberName).thenAccept(__ -> topicSubscriptions.remove(topic));
+            return CompletableFuture.completedFuture(null);
         }
     }
 
@@ -75,11 +81,6 @@ public class TopicSubscriptionManager {
                 .map(topic -> unsubscribe(topic, true))
                 .collect(Collectors.toList());
         return FutureUtil.waitForAll(futures);
-    }
-
-    @VisibleForTesting
-    public Map<Topic, Pair<Subscription, Consumer>> getTopicSubscriptions() {
-        return topicSubscriptions;
     }
 
     private void removeConsumerIfExist(Subscription subscription, Consumer consumer) throws BrokerServiceException {

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/TopicSubscriptionManager.java
@@ -20,7 +20,6 @@ import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.commons.collections4.CollectionUtils;
 import org.apache.commons.lang3.tuple.Pair;
 import org.apache.pulsar.broker.service.BrokerServiceException;
 import org.apache.pulsar.broker.service.Consumer;
@@ -61,15 +60,9 @@ public class TopicSubscriptionManager {
             log.error("[ Subscription ] Subscription {} Remove consumer fail.", subscriberName, e);
             FutureUtil.failedFuture(e);
         }
-        if (CollectionUtils.isEmpty(subscriptionConsumerPair.getLeft().getConsumers())) {
-            return topic.unsubscribe(subscriberName)
-                    .thenCompose(__ -> {
-                        if (cleanSubscription) {
-                            return subscriptionConsumerPair.getLeft().deleteForcefully()
-                                    .thenAccept(unused -> topicSubscriptions.remove(topic));
-                        }
-                        return CompletableFuture.completedFuture(null);
-                    });
+        if (cleanSubscription) {
+            return subscriptionConsumerPair.getLeft().deleteForcefully()
+                    .thenAccept(unused -> topicSubscriptions.remove(topic));
         } else {
             return CompletableFuture.completedFuture(null);
         }

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
@@ -215,7 +215,6 @@ public class MQTTProxyProtocolMethodProcessor extends AbstractCommonProtocolMeth
                 log.debug("[Proxy Connection Lost] [{}] ", connection.getClientId());
             }
             connectionManager.removeConnection(connection);
-            connection.close();
         }
         brokerPool.forEach((k, v) -> v.close());
         brokerPool.clear();

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java
@@ -203,8 +203,8 @@ public class MQTTProxyProtocolMethodProcessor extends AbstractCommonProtocolMeth
             log.warn("connection is null. close CId={}", clientId);
             channel.close();
         } else {
-            connection.close()
-                    .thenAccept(__-> connectionManager.removeConnection(connection));
+            connectionManager.removeConnection(connection);
+            connection.close();
         }
     }
 
@@ -215,6 +215,7 @@ public class MQTTProxyProtocolMethodProcessor extends AbstractCommonProtocolMeth
                 log.debug("[Proxy Connection Lost] [{}] ", connection.getClientId());
             }
             connectionManager.removeConnection(connection);
+            connection.close();
         }
         brokerPool.forEach((k, v) -> v.close());
         brokerPool.clear();

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
@@ -211,7 +211,8 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                 return this.qosPublishHandlers.qos2().publish(msg);
             default:
                 log.error("[Publish] Unknown QoS-Type:{}", qos);
-                return connection.close();
+                connection.getChannel().close();
+                return FutureUtil.failedFuture(new IllegalArgumentException("Unknown Qos-Type: " + qos));
         }
     }
 
@@ -269,11 +270,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
                     .success(true)
                     .build();
             connection.getAckHandler()
-                    .sendDisconnectAck(connection, disconnectAck)
-                    .addListener(__ -> {
-                        metricsCollector.removeClient(NettyUtils.getAddress(channel));
-                        connectionManager.removeConnection(connection);
-                    });
+                    .sendDisconnectAck(connection, disconnectAck);
         }
     }
 
@@ -292,11 +289,9 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
         if (willMessage != null) {
             willMessageHandler.fireWillMessage(clientId, willMessage);
         }
-        connection.close()
-                .thenAccept(__ -> {
-                    connectionManager.removeConnection(connection);
-                    mqttSubscriptionManager.removeSubscription(clientId);
-                });
+        connectionManager.removeConnection(connection);
+        mqttSubscriptionManager.removeSubscription(clientId);
+        connection.close();
     }
 
     @Override

--- a/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
+++ b/mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/DefaultProtocolMethodProcessorImpl.java
@@ -268,6 +268,7 @@ public class DefaultProtocolMethodProcessorImpl extends AbstractCommonProtocolMe
             DisconnectAck disconnectAck = DisconnectAck
                     .builder()
                     .success(true)
+                    .reasonCode(Mqtt5DisConnReasonCode.NORMAL)
                     .build();
             connection.getAckHandler()
                     .sendDisconnectAck(connection, disconnectAck);

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
@@ -46,6 +46,7 @@ import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.client.api.Schema;
 import org.apache.pulsar.common.naming.TopicDomain;
 import org.apache.pulsar.common.policies.data.BundlesData;
+import org.apache.pulsar.common.policies.data.TopicStats;
 import org.awaitility.Awaitility;
 import org.fusesource.mqtt.client.BlockingConnection;
 import org.fusesource.mqtt.client.MQTT;
@@ -293,6 +294,10 @@ public class SimpleIntegrationTest extends MQTTTestBase {
         connection2.subscribe(topics);
         Assert.assertTrue(connection2.isConnected());
         connection2.disconnect();
+        Awaitility.await().untilAsserted(()-> {
+            TopicStats stats = admin.topics().getStats(topicName);
+            Assert.assertEquals(stats.getSubscriptions().size(), 0);
+        });
         connection1.disconnect();
     }
 

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
@@ -270,7 +270,7 @@ public class SimpleIntegrationTest extends MQTTTestBase {
         connection.disconnect();
     }
 
-    @Test(timeOut = TIMEOUT, invocationCount = 1000)
+    @Test(timeOut = TIMEOUT)
     public void testSubscribeRejectionWithSameClientId() throws Exception {
         final String topicName = "persistent://public/default/testSubscribeWithSameClientId";
         MQTT mqtt = createMQTTClient();

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt3/fusesource/base/SimpleIntegrationTest.java
@@ -270,11 +270,13 @@ public class SimpleIntegrationTest extends MQTTTestBase {
         connection.disconnect();
     }
 
-    @Test(timeOut = TIMEOUT)
+    @Test(timeOut = TIMEOUT, invocationCount = 1000)
     public void testSubscribeRejectionWithSameClientId() throws Exception {
         final String topicName = "persistent://public/default/testSubscribeWithSameClientId";
         MQTT mqtt = createMQTTClient();
         mqtt.setClientId("client-id-0");
+        mqtt.setReconnectDelay(Integer.MAX_VALUE);
+        mqtt.setReconnectAttemptsMax(0);
         BlockingConnection connection1 = mqtt.blockingConnection();
         connection1.connect();
         Topic[] topics = { new Topic(topicName, QoS.AT_LEAST_ONCE) };
@@ -283,12 +285,13 @@ public class SimpleIntegrationTest extends MQTTTestBase {
 
         BlockingConnection connection2;
         MQTT mqtt2 = createMQTTClient();
-        mqtt.setClientId("client-id-0");
+        mqtt2.setClientId("client-id-0");
         connection2 = mqtt2.blockingConnection();
         connection2.connect();
         Assert.assertTrue(connection2.isConnected());
-        Awaitility.await().untilAsserted(() -> Assert.assertTrue(connection1.isConnected()));
+        Awaitility.await().untilAsserted(() -> Assert.assertFalse(connection1.isConnected()));
         connection2.subscribe(topics);
+        Assert.assertTrue(connection2.isConnected());
         connection2.disconnect();
         connection1.disconnect();
     }

--- a/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5ReasonCodeOnAllACKTest.java
+++ b/tests/src/test/java/io/streamnative/pulsar/handlers/mqtt/mqtt5/hivemq/base/MQTT5ReasonCodeOnAllACKTest.java
@@ -107,13 +107,14 @@ public class MQTT5ReasonCodeOnAllACKTest extends MQTTTestBase {
         }
         client.unsubscribeWith().topicFilter(topic).send();
         client.disconnect();
-        client.connect();
-        Mqtt5SubAck secondConsumerACK = client.subscribeWith().topicFilter(topic).qos(MqttQos.AT_MOST_ONCE).send();
+        Mqtt5BlockingClient client2 = MQTT5ClientUtils.createMqtt5Client(getMqttBrokerPortList().get(0));
+        client2.connect();
+        Mqtt5SubAck secondConsumerACK = client2.subscribeWith().topicFilter(topic).qos(MqttQos.AT_MOST_ONCE).send();
         for (Mqtt5SubAckReasonCode reasonCode : secondConsumerACK.getReasonCodes()) {
             Assert.assertEquals(reasonCode.getCode(), Mqtt5SubAckReasonCode.GRANTED_QOS_0.getCode());
         }
-        client.unsubscribeWith().topicFilter(topic).send();
-        client.disconnect();
+        client2.unsubscribeWith().topicFilter(topic).send();
+        client2.disconnect();
     }
 
     @Test(dataProvider = "mqttPersistentTopicNames", timeOut = TIMEOUT)

